[OpenMP][Clang] Force use of `num_teams` and `thread_limit` for bare kernel

shiltian opened the pull request:
This patch makes `num_teams` and `thread_limit` mandatory for bare kernels,
similar to a reguar kernel language that when launching a kernel, the grid size
has to be set explicitly.

jhuber6 commented:
Does `thread_limit` directly imply the number of threads? I thought that it merely set an upper bound such that it cannot be increased beyond that via environment variables.

shiltian commented:
> Does `thread_limit` directly imply the number of threads? I thought that it merely set an upper bound such that it cannot be increased beyond that via environment variables.

Based on the spec, yes. However, here since `ompx_bare` is an extension, we can redefine semantics. For example, we already redefine the `target teams` region such that globalization is disabled. We can say, `thread_limit` in this mode is to set the block size. I'll have a follow-up patch to change the runtime behavior that if user's grid size can not be met in such kernel mode, crash directly, similar to CUDA/HIP.

jdoerfert commented:
I think the follow up, to force the user bound for bare kernels, make sense. I am not sold on this patch though. Why would we disallow users to do the same looping we do in the deviceRTL while hoping the offload runtime will pick a good grid size?

shiltian commented:
> I think the follow up, to force the user bound for bare kernels, make sense. I am not sold on this patch though. Why would we disallow users to do the same looping we do in the deviceRTL while hoping the offload runtime will pick a good grid size?

Because we don't have loop trip count in this case, so the runtime picks how many, 3200 thread blocks and 128 threads per thread block IIRC. I'm not sure that can be called a "good" grid size and we don't have any heuristic w/o loop trip count anyway.

Typically when writing a CUDA/HIP kernel, users calculate the grid/block size manually and launch the kernel using that sizes. That is the main reason for this patch. This can also make the runtime decision much easier: if we can't meet users' requirement, we crash.

shiltian commented:
Gentle ping

alexey-bataev commented on clang/lib/Sema/SemaOpenMP.cpp:
```diff
@@ -14633,6 +14633,26 @@ StmtResult Sema::ActOnOpenMPTargetTeamsDirective(ArrayRef<OMPClause *> Clauses,
+      BareClause = C;
+      HasBareClause = true;
+    } else if (isa<OMPNumTeamsClause>(C))
+      HasNumTeamsClause = true;
```
Enclose all substatements in to braces to make it uniform

alexey-bataev commented on clang/lib/Sema/SemaOpenMP.cpp:
```diff
@@ -14633,6 +14633,26 @@ StmtResult Sema::ActOnOpenMPTargetTeamsDirective(ArrayRef<OMPClause *> Clauses,
+  for (auto *C : Clauses) {
+    if (isa<OMPXBareClause>(C)) {
+      BareClause = C;
+      HasBareClause = true;
```
Do you really need this flag? To just check that BareClause != nullptr is enough?

shiltian replied:
It is not very necessarily needed, but I found it look nicer to check three boolean variables instead of two as well as a pointer. Just code preference. Nothing technical.

shiltian committed:
[OpenMP][Clang] Force use of `num_teams` and `thread_limit` for bare kernel

This patch makes `num_teams` and `thread_limit` mandatory for bare kernels,
similar to a reguar kernel language that when launching a kernel, the grid size
has to be set explicitly.

shiltian commented:
gentle ping

alexey-bataev commented:
Sorry, on a vacation, will review on Monday
